app/src/main/java/com/example/android/codelabs/paging/data/GithubPagingSource.kt

```diff
@@ -0,0 +1,29 @@
+package com.example.android.codelabs.paging.data
+
+import androidx.paging.PagingSource
+import com.example.android.codelabs.paging.api.GithubService
+import com.example.android.codelabs.paging.api.searchRepos
+import com.example.android.codelabs.paging.model.Repo
+import com.example.android.codelabs.paging.model.RepoSearchResult
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.FlowPreview
+
+@ExperimentalCoroutinesApi
+@FlowPreview
+class GithubPagingSource(
+        private val service: GithubService,
+        private val query: String
+) : PagingSource<Int, Repo>() {
+    override suspend fun load(params: LoadParams<Int>): LoadResult<Int, Repo> {
+        val position = params.key ?: 0
+        val apiResponse = searchRepos(service, query, position, GithubRepository.NETWORK_PAGE_SIZE)
+        return when(apiResponse){
+            is RepoSearchResult.Success -> LoadResult.Page(
+                    data = apiResponse.data,
+                    prevKey = position,
+                    nextKey = position + 1
```

Comment:
same as above, you need to pass null to terminate.

can the response tell you whether there's more data left?

Reply:
using how much data I got to determine if i'm at the last page. I don't think i can do better than this.